src/platform/FailSafeContext.cpp

```diff
@@ -37,36 +37,32 @@ constexpr TLV::Tag kUpdateNocCommandTag = TLV::ContextTag(2);
 
 void FailSafeContext::HandleArmFailSafeTimer(System::Layer * layer, void * aAppState)
 {
-    FailSafeContext * context = reinterpret_cast<FailSafeContext *>(aAppState);
-    context->FailSafeTimerExpired();
+    FailSafeContext * failSafeContext = reinterpret_cast<FailSafeContext *>(aAppState);
+    failSafeContext->FailSafeTimerExpired();
 }
 
 void FailSafeContext::HandleDisarmFailSafe(intptr_t arg)
 {
-    FailSafeContext * this_ = reinterpret_cast<FailSafeContext *>(arg);
-
-    this_->mFailSafeBusy = false;
-
-    if (ConfigurationMgr().SetFailSafeArmed(false) != CHIP_NO_ERROR)
-    {
-        ChipLogError(DeviceLayer, "Failed to set FailSafeArmed config to false");
-    }
-
-    if (DeleteFromStorage() != CHIP_NO_ERROR)
-    {
-        ChipLogError(DeviceLayer, "Failed to delete FailSafeContext from configuration");
-    }
+    FailSafeContext * failSafeContext = reinterpret_cast<FailSafeContext *>(arg);
+    failSafeContext->DisarmFailSafe();
 }
 
 void FailSafeContext::FailSafeTimerExpired()
 {
+    if (!IsFailSafeArmed())
+    {
+        // In case this was a pending timer event in event loop, and we had
+        // done CommissioningComplete or manual disarm.
+        return;
+    }
+
+    ChipLogProgress(FailSafe, "Fail-safe timer expired");
+    DeviceLayer::SystemLayer().CancelTimer(HandleArmFailSafeTimer, this);
```

Comment:
I don't understand this CancelTimer call.  The timer just fired.... why do we need to cancel it again?

Reply:
Fixed